src/solver/conv_asm_1x1u.cpp

```diff
@@ -307,11 +354,150 @@ bool PerformanceConfigConvAsm1x1U::IsValid(const ProblemDescription& problem) co
+            return false;
+        if(problem.direction.IsBackwardData() && !(problem.n_outputs % k_mult == 0))
+            return false;
+    }
+    if(chunks_per_wave != -1)
+    {
+        if(!(read_size * elements_in_dword <= chunks_per_wave))
+            return false;
+        if(chunks_per_wave % elements_in_dword != 0)
+            return false;
+    }
+    if(chunk_size != -1)
+    {
+        const int total_chunks = (img_hw + chunk_size - 1) / chunk_size;
+        if(!(chunks_per_wave <= total_chunks))
+            return false;
+    }
+    if(n_mult != -1)
+    {
+        const int total_n_blocks = (problem.batch_sz + GetNPerGpr() - 1) / GetNPerGpr();
+        if(!(n_mult <= total_n_blocks))
+            return false;
+    }
+    const auto in_gprs =
+        (chunks_per_wave * n_mult * c_mult + elements_in_dword - 1) / elements_in_dword;
+    const auto acc_gprs = chunks_per_wave * n_mult * k_mult;
+    // TODO last vgpr only for old card.
+    // ADD if(option.machine_version_major == 9)
+    // vgprs  = 4 + 2 * in_gprs + acc_gprs + (img_hw % elements_in_dword != 0 ? 1: 0);
+    // else
+    const auto vgprs = 4 + 2 * in_gprs + acc_gprs + (img_hw % elements_in_dword != 0 ? 1 : 0) + 1;
+    if(c_mult != -1)
+    {
+        if((c_mult % elements_in_dword) != 0)
+            return false;
+        if(!(vgprs < 256))
+            return false;
+        const auto sgprs = 25 + 2 * k_mult * c_mult;
+        if(!(sgprs < 102)) /// \todo This is valid for Gfx8 and Gfx9. Check for newer parts.
+            return false;
+    }
+    if(waves_c_in_group != -1)
+    {
+        if(!(waves_c_in_group <= problem.n_inputs))
+            return false;
+        const int c_per_wave      = (problem.n_inputs + waves_c_in_group - 1) / waves_c_in_group;
+        const int c_per_last_wave = problem.n_inputs - (c_per_wave * (waves_c_in_group - 1));
+        if(c_per_wave % c_mult != 0 || c_per_last_wave % c_mult != 0)
+            return false;
+        // return (c_per_wave % c_mult == 0) && (c_per_last_wave % c_mult == 0);
+    }
+    if(waves_k_in_group != -1)
+    {
+        if(!(k_mult * waves_k_in_group <= problem.n_outputs))
+            return false;
+        if(!(waves_c_in_group * waves_k_in_group <= 16))
+            return false;
+        const auto max_waves_per_CU = (256 / vgprs) * 4;
+        if(!(max_waves_per_CU >= waves_c_in_group * waves_k_in_group))
+            return false;
+    }
+    return true;
+}
+
+bool PerformanceConfigConvAsm1x1U::TryToken(int index, int value, const ProblemDescription& problem)
+{
+    switch(index)
+    {
+    case 0: read_size = value; break;
+    case 1: k_mult = value; break;
+    case 2: chunks_per_wave = value; break;
+    case 3: chunk_size = value; break;
+    case 4: n_mult = value; break;
+    case 5: c_mult = value; break;
+    case 6: waves_c_in_group = value; break;
+    case 7: waves_k_in_group = value; break;
+    default: return false;
+    }
+    return this->IsValidDynamic(problem);
```

Comment:
I thought the same thing initially as well, but if it does try a token and fail, a new token will be tried. If all tries fail, then HeuristicInit() will fall back to the default heuristic which will set it. Also if the default heuristic fails it does not reset all tokens to -1, this is because -1 is about as valid as any other value.

Reply:
@Dmantri98 If this function may leave the data in invalid state, then this must be clearly documented in comments.

Reply:
Then it should be named like ApplyToken, because it actually applies it no matter what, even if it would not be valid. Otherwise, the reader has to look into the implementation carefully to understand what it does. In general, TrySomeAction methods usually do not change the state when they return failure.

Reply:
@DrizztDoUrden Right. 

@Dmantri98 [Recommendation]:
- As we do not have namespaces within classes, please change the name so it becomes clear that this is AI tuning related. Thanks.
- Ditto wrt `RunParmeterPredictionModel`

Reply:
"Model" usually entails that it is some sort of Artificial Intelligence/Neural Network.

Reply:
Also both of these functions are inside this cmake check:
#if MIOPEN_ENABLE_AI_KERNEL_TUNING.
Nevertheless I can change this to ModelApplyToken().

Reply:
@Dmantri98 
> "Model" usually entails that it is some sort of Artificial Intelligence/Neural Network.

We already have models that compute WTI for certain kernels.

> Also both of these functions are inside this cmake check:
#if MIOPEN_ENABLE_AI_KERNEL_TUNING.

Indeed.

> Nevertheless I can change this to ModelApplyToken().

Well, if you decide to rename, then I would recommend using the same approach _as if_ we had namespaces in classes, i.e. adding some common prefix to methods related to AI tuning.

But all this is up to you -- [my comment](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2031#discussion_r1158504361) is just a recommendation.